### PR TITLE
Issue #873 quote previous reply

### PR DESCRIPTION
**Pull request by tomholub**
I had to create a PR so that I can make a review with comments, it's clearer that way
#873

### CODE REVIEW

File: extension/chrome/elements/compose.ts [hunk 1]
```diff
@@ -83,7 +83,7 @@ Catch.try(async () => {
     draftId = storage.drafts_reply[threadId]; // there may be a draft we want to load
   }
 
-  const processedUrlParams = { acctEmail, draftId, threadId, subject, from, to, frameId, tabId, isReplyBox, skipClickPrompt, parentTabId, disableDraftSaving, debug };
+  const processedUrlParams = { acctEmail, draftId, threadId, threadMsgId, subject, from, to, frameId, tabId, isReplyBox, skipClickPrompt, parentTabId, disableDraftSaving, debug };
```
Comment (tomholub) [inline extension/chrome/elements/compose.ts]: The threadMsgId is used as a backup mechanism in case we could not parse threadId. In fact I want to remove & deprecate its use, but didn't have a chance to do it yet.

The issue is that it's not guaranteed to be there, you should go by threadId.

File: extension/chrome/elements/reply_pubkey_mismatch.ts [hunk 2]
```diff
@@ -23,6 +23,7 @@ Catch.try(async () => {
   const subject = Assert.urlParamRequire.optionalString(uncheckedUrlParams, 'subject') || '';
   const frameId = Assert.urlParamRequire.string(uncheckedUrlParams, 'frameId');
   const threadId = Assert.urlParamRequire.optionalString(uncheckedUrlParams, 'threadId') || '';
+  const threadMsgId = Assert.urlParamRequire.optionalString(uncheckedUrlParams, 'threadMsgId') || '';
```
Comment (tomholub) [inline extension/chrome/elements/reply_pubkey_mismatch.ts]: Sorry - as above - better remove.

File: extension/chrome/elements/reply_pubkey_mismatch.ts [hunk 3]
```diff
@@ -32,7 +33,7 @@ Catch.try(async () => {
   const appFunctions = Composer.defaultAppFunctions();
   const tabId = await BrowserMsg.requiredTabId();
   const processedUrlParams = {
-    acctEmail, draftId: '', threadId, subject, from, to, frameId, tabId, debug,
+    acctEmail, draftId: '', threadId, subject, from, to, frameId, threadMsgId, tabId, debug,
```
Comment (tomholub) [inline extension/chrome/elements/reply_pubkey_mismatch.ts]: Sorry - as above - better remove.

File: extension/js/common/composer.ts [hunk 4]
```diff
@@ -67,6 +67,7 @@ export type ComposerUrlParams = {
   tabId: string;
   acctEmail: string;
   threadId: string;
+  threadMsgId: string;
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Sorry - as above - better remove.

File: extension/js/common/composer.ts [hunk 5]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
         if (method === 'forward') {
           this.urlParams.subject = 'Fwd: ' + this.urlParams.subject;
           await this.retrieveDecryptAddForwardedMsg(determined.lastMsgId);
+        } else {
+          try {
+            const { raw } = await Google.gmail.msgGet(this.urlParams.acctEmail, this.urlParams.threadMsgId, 'raw');
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Although `threadMsgId` usually points to the last message, this is not guaranteed and I'd like to stop passing it into compose one day.

You can use `threadId` to pull a list of messages, then grab the last one. Please look at what forward is doing, and re-use (some of) the code.

File: extension/js/common/composer.ts [hunk 6]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+            const msgMimeContent = await Mime.process(Buf.fromBase64UrlStr(raw!));
+            if (!msgMimeContent.from || !msgMimeContent.headers.date || !msgMimeContent.blocks || !msgMimeContent.blocks.length) {
+              return;
```
Comment (tomholub) [inline extension/js/common/composer.ts]: This is a fair precaution, though I cannot imagine it would actually happen. Maybe indeed `!msgMimeContent.blocks.length` could be true for empty messages, not sure.

But it's ok to check, just some thoughts.

File: extension/js/common/composer.ts [hunk 7]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+            }
+            if (!this.urlParams.draftId) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: This took me a while to realize the reason for the check. Please add a comment
```js
// if there is a draftId, then the quoted message was already pulled before and added into the draft
```

File: extension/js/common/composer.ts [hunk 8]
```diff
@@ -1827,6 +1873,24 @@ export class Composer {
     }
   }
 
+  private quoteText(text: string) {
+    return text.split('\n').map(l => '<br/>&gt; ' + l).join('\n');
+  }
+
+  private formatDate(date: Date) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: maybe shorten to 
```js
date.toISOString().replace(/T/, ' ').replace(/:[^:]+$/, '')
// "2019-07-02 08:15"
```
Then you could add a method as `Str.fromDate` or something like that.

File: extension/js/common/core/mime.ts [hunk 9]
```diff
@@ -15,7 +15,7 @@ const MimeBuilder = requireMimeBuilder();  // tslint:disable-line:variable-name
 const Iso88592 = requireIso88592();  // tslint:disable-line:variable-name
 
 type MimeContentHeader = string | { address: string; name: string; }[];
-type MimeContent = {
+export type MimeContent = {
```
Comment (tomholub) [inline extension/js/common/core/mime.ts]: It's not clear that the export is needed, at least based on the code above. It seems it would work without the export.

Nothing wrong with exporting, but if it's not used elsewhere, better keep it local, so that it's easier to discover when it's unused (local unused variable goes gray, exported does not)

File: extension/js/common/composer.ts [hunk 10]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+          try {
+            const { raw } = await Google.gmail.msgGet(this.urlParams.acctEmail, this.urlParams.threadMsgId, 'raw');
+            const msgMimeContent = await Mime.process(Buf.fromBase64UrlStr(raw!));
+            if (!msgMimeContent.from || !msgMimeContent.headers.date || !msgMimeContent.blocks || !msgMimeContent.blocks.length) {
+              return;
+            }
+            if (!this.urlParams.draftId) {
+              const blocksToDisplay = msgMimeContent.blocks.filter(b => b.type === 'encryptedMsg' ||
+                b.type === 'plainText' || b.type === 'plainHtml');
+              const contentToDisplay: Array<string> = [];
```
Comment (tomholub) [inline extension/js/common/composer.ts]: All of this below looks good, but it's somewhat duplicated. Very similar functionality is in `retrieveDecryptAddForwardedMsg`, so these two should be merged together and put into one function (the pulling and decrypting of the message, which just returns plain text).

I think your method does it slightly better, because you are cycling through all of the blocks with `Mime.process`, so you'll be sure to catch the needed content.

On the other hand, you are pulling the raw message. If it contains a 15 mb attachment, that may take a while. So you may need to add a loading % sign, maybe next to the three dots, eg "Loading previous message quote... 18%". That means the three dots should probably be visible while this happens, but have opacity 50% and clicking it should not do anything (maybe should also have `cursor: default`). Then when it loads, opacity of (...) goes to 1 and it can be clicked if user wants to.

Either that, or adopt the faster but less precise method from forwarding messages, which may just be good enough. I don't know, your call - you can test it on `flowcrypt.compatibility` emails to see how many of them (that are not marked with a `fix` label) load without issues. You should be able to load almost any plain, plain html, encrypted, or signed message.

File: extension/js/common/composer.ts [hunk 11]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+              for (const block of blocksToDisplay) {
+                if (block.type === 'encryptedMsg') {
+                  const armored = Pgp.armor.clip(block.content as string);
+                  if (armored) {
+                    const decrypted = await PgpMsg.decrypt({
+                      kisWithPp: await Store.keysGetAllWithPassphrases(this.urlParams.acctEmail),
+                      encryptedData: Buf.fromUtfStr(armored)
+                    });
+                    if (decrypted.success) {
+                      contentToDisplay.push(decrypted.content.toUtfStr());
```
Comment (tomholub) [inline extension/js/common/composer.ts]:
```js
} else {
  contentToDisplay.push(`\n(Failed to decrypt quote from previous message because: ${decrypted.error.type}: ${decrypted.error.message})\n`);
}
```
or something like that. Easier to debug issues

File: extension/js/common/composer.ts [hunk 12]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+            if (!msgMimeContent.from || !msgMimeContent.headers.date || !msgMimeContent.blocks || !msgMimeContent.blocks.length) {
+              return;
+            }
+            if (!this.urlParams.draftId) {
+              const blocksToDisplay = msgMimeContent.blocks.filter(b => b.type === 'encryptedMsg' ||
+                b.type === 'plainText' || b.type === 'plainHtml');
+              const contentToDisplay: Array<string> = [];
+              for (const block of blocksToDisplay) {
+                if (block.type === 'encryptedMsg') {
+                  const armored = Pgp.armor.clip(block.content as string);
```
Comment (tomholub) [inline extension/js/common/composer.ts]: You're already getting and `encryptedMsg` block, it should no longer need clipping.

File: extension/js/common/composer.ts [hunk 13]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+                  if (armored) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Here the if should not be needed.

File: extension/js/common/composer.ts [hunk 14]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+                    const decrypted = await PgpMsg.decrypt({
+                      kisWithPp: await Store.keysGetAllWithPassphrases(this.urlParams.acctEmail),
+                      encryptedData: Buf.fromUtfStr(armored)
+                    });
+                    if (decrypted.success) {
+                      contentToDisplay.push(decrypted.content.toUtfStr());
+                    }
+                  }
+                } else if (block.type === 'plainHtml') {
+                  contentToDisplay.push(Xss.escape(block.content as string));
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Escaping would result in ugly html tags in the resulting quote text, I think. Can do `Xss.htmlSanitizeAndStripAllTags(content, '\n')` so that you actually convert the html to text, with proper (guessed) newlines.

Let me know if I misunderstood

File: extension/js/common/composer.ts [hunk 15]
```diff
@@ -1201,6 +1208,45 @@ export class Composer {
...
+                } else {
+                  contentToDisplay.push(block.content as string);
+                }
+              }
+              const sentDate = new Date(msgMimeContent.headers.date as string);
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Safer to do `String(x)` (if the type is `String | other things`) or alternatively `x || ''` (if the type is `string | undefined`) rather than casting it, when possible / appropriate.

File: extension/js/common/composer.ts (hunk 13 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Pgp.armor.clip returns `string | undefined`, because of this, I added `if` expression

File: extension/js/common/composer.ts (hunk 14 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Okay, thanks. `Xss.htmlSanitizeAndStripAllTags(content, '\n')` would be better here

File: extension/js/common/composer.ts (hunk 15 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Got it, Thanks!

File: extension/js/common/composer.ts (hunk 5 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Okay, I will use threadId to get a list and then pick up the last message from the list

File: extension/chrome/elements/compose.ts (hunk 1 above)
Comment (michael-volynets) [reply extension/chrome/elements/compose.ts]: Okay, I will remove `threadMsgId`. Thanks!

File: extension/js/common/composer.ts (hunk 13 above)
Comment (tomholub) [reply extension/js/common/composer.ts]: I see it now. Once you remove the clip, you can remove this too.

File: extension/js/common/core/mime.ts (hunk 9 above)
Comment (michael-volynets) [reply extension/js/common/core/mime.ts]: oh, I forget to remove it,
I tried to use `Mime.decode` before but I didn' work well and I removed it, and forget to remove `export`.
Thanks

File: extension/chrome/elements/compose.ts (hunk 1 above)
Comment (michael-volynets) [reply extension/chrome/elements/compose.ts]: in the code when we forward a message I can see you get last msgId using `await this.app.emailProviderDetermineReplyMsgHeaderVariables();`.
I think I can use this one too instead of getting all the messages in the thread. Am I correct?

Comment (tomholub) [reply extension/chrome/elements/compose.ts]: Yes, you can do it similarly to what forward does.

**Comment by michael-volynets** [comment]
@tomholub I almost did it, it quotes replied message but it works laggy because of loading message and decrypting it. I need to think about how can I fix it.
So you can review it but don't merge it yet

**Comment by tomholub** [comment]
You may need to add a % loader next to it, since it also needs to download the message, which can be big and take time. You could just show loader up to 60% (= 100% dowloaded), and then show it stuck at 60 until decrypting is done too.

I also saw that there was still the large date formatting function, if possible please replace.

**Comment by michael-volynets** [comment]
I also thought about loaders. 
I think they can fix the problem but I think would be better to start loading the message to composer even user didn't click the reply button and when they click on the button (if they click) the replied part might show immediately. The minus of this way that the user might not click to the reply button but the request for the message will be sent.

**Comment by tomholub** [comment]
Some messages can be up to 35mb if you are pulling the raw message. That's a lot of things to pull.

**Comment by michael-volynets** [comment]
Wow, okay, then that's a bad idea to pull it without clicking on reply button

File: extension/js/common/composer.ts [hunk 16]
```diff
@@ -1149,58 +1163,43 @@ export class Composer {
...
-          await this.retrieveDecryptAddForwardedMsg(determined.lastMsgId);
+          await this.appendForwardedMsg(this.messageToReplyOrForward.text);
+        } else {
+          if (!this.messageToReplyOrForward.headers.from || !this.messageToReplyOrForward.headers.date) {
+            return;
+          }
+          const sentDate = new Date(String(this.messageToReplyOrForward.headers.date));
+          this.msgExpandingHTMLPart = '<br><br>' + this.generateHTMLRepliedPart(this.messageToReplyOrForward.text, sentDate, this.messageToReplyOrForward.headers.from);
+          if (!this.urlParams.draftId) {
+            this.showExpandButton(this.msgExpandingHTMLPart);
```
Comment (tomholub) [inline extension/js/common/composer.ts]: No need to resize compose box here?
(I'm not sure)

File: extension/js/common/composer.ts [hunk 17]
```diff
@@ -1653,8 +1714,8 @@ export class Composer {
     }, this.getErrHandlers(`focus on recipient field`))).children().click(() => false);
     this.resizeInputTo();
     this.attach.initAttDialog('fineuploader', 'fineuploader_button');
-    this.attach.setAttAddedCb(async () => this.setInputTextHeightManuallyIfNeeded());
-    this.attach.setAttRemovedCb(() => this.setInputTextHeightManuallyIfNeeded());
+    this.attach.setAttAddedCb(async () => { this.setInputTextHeightManuallyIfNeeded(); this.resizeComposeBox(); });
```
Comment (tomholub) [inline extension/js/common/composer.ts]: when there is more than one command, please format it as multiline (max one statement; per line).

File: extension/js/common/composer.ts [hunk 18]
```diff
@@ -1653,8 +1714,8 @@ export class Composer {
...
+    this.attach.setAttRemovedCb(() => { this.setInputTextHeightManuallyIfNeeded(); this.resizeComposeBox(); });
```
Comment (tomholub) [inline extension/js/common/composer.ts]: format as multiline block

File: extension/js/common/core/mime.ts [hunk 19]
```diff
@@ -42,10 +42,16 @@ export type MsgBlock = {
   attMeta?: AttMeta; // only in plainAtt, encryptedAtt, decryptedAtt, encryptedAttLink (not sure if always)
   decryptErr?: DecryptError; // only in decryptErr block, always
 };
+export type MimeProccesedMsg = {
```
Comment (tomholub) [inline extension/js/common/core/mime.ts]: Yes, why not. Thanks

File: extension/js/content_scripts/webmail/gmail_element_replacer.ts [hunk 20]
```diff
@@ -125,11 +125,7 @@ export class GmailElementReplacer implements WebmailElementReplacer {
         const lastReplyBtn = $(convoReplyBtnsToReplaceArr.pop()!);
         $(lastReplyBtn).addClass('inserted');
         const element = $(this.factory.btnReply()).insertBefore($(lastReplyBtn).children().last());
-        if (isEncrypted) {
-          element.click(Ui.event.prevent('double', Catch.try(this.setReplyBoxEditable)));
-        } else {
-          element.click(Ui.event.prevent('double', Catch.try(() => this.replaceStandardReplyBox(true, true)))); // xss-safe-factory
-        }
+        element.click(Ui.event.prevent('double', Catch.try(this.setReplyBoxEditable)));
```
Comment (tomholub) [inline extension/js/content_scripts/webmail/gmail_element_replacer.ts]: Why this change? I think the `true, true` is a way to force replacement of the standard reply box even when the conversation is otherwise not encrypted (eg clicking "encrypted reply") on a regular conversation.

**Comment by tomholub** [comment]
When this is all done, and we have added tests for the behavior, we'll need to refactor it. The composer file is getting too big. I'm thinking all draft functionality will go into some `ComposerDraft` class, all quoting functionality would go into a `ComposerQuote` class, and finally a `ComposerContacts` class. I think that covers a big chunk of the functionality.

Don't do it yet, until we test all this more thoroughly.

**Comment by tomholub** [comment]
That will be fore another issue / another PR.

File: extension/js/common/composer.ts (hunk 16 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: resizeComposeBox is below, after `else` statement

File: extension/js/common/composer.ts (hunk 17 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Okay, got it

File: extension/js/common/composer.ts (hunk 16 above)
Comment (tomholub) [reply extension/js/common/composer.ts]: Right, but that one will only run if you enter the `else` branch. How about in this `if` branch evaluates true?

**Comment by michael-volynets** [comment]
Issue # 873

Time Spent: ~ 2d 4h
There were some issues, like hiding quoted part when a draft is loaded or when the message is loading very long.
I think I could have made it faster because sometimes I removed the code that I had written because I started writing in the incorrect way from the beginning but in general I've obtained a lot of experience with the composer by working on this task. 
Thanks

File: extension/js/common/composer.ts [hunk 21]
```diff
@@ -1383,21 +1390,26 @@ export class Composer {
...
-        const { raw } = await Google.gmail.msgGet(this.urlParams.acctEmail, determined.lastMsgId, 'raw');
+        const { raw } = await Google.gmail.msgGet(this.urlParams.acctEmail, determined.lastMsgId, 'raw',
+          (progress: number) => this.renderGettingPreviousMessageProgress(30 + progress * 0.3));
         const message = await Mime.process(Buf.fromBase64UrlStr(raw!));
         const readableBlocks = message.blocks
           .filter(b => b.type === 'encryptedMsg' || b.type === 'plainText' || b.type === 'plainHtml');
+        const encryptedCount = readableBlocks.filter(b => b.type === 'encryptedMsg').length;
         const decryptedAndFormatedContent: string[] = [];
-        for (const block of readableBlocks) {
+        for (const [index, block] of readableBlocks.entries()) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Interesting, I haven't seen that before. That may come handy. I usually did a traditional for loop with i++, but this may be better if I really need both.

**Comment by tomholub** [comment]
Re-running tests

File: extension/js/common/composer.ts (hunk 21 above)
Comment (michael-volynets) [reply extension/js/common/composer.ts]: Honestly, I'm using this first time :) 
I don't like traditional `for` statement because it's hard to read.

I'm 100% sure that traditional `for` would be better for productivity but the difference won't be very big (less the 1ms I think).

Comment (tomholub) [reply extension/js/common/composer.ts]: I'm pretty sure the performance difference is meaningless

File: extension/js/common/api/google.ts [hunk 22]
```diff
@@ -145,9 +145,10 @@ export class Google extends EmailProviderApi {
       r.emailAddress = r.emailAddress.toLowerCase();
       return r;
     },
-    threadGet: (acctEmail: string, threadId: string, format?: GmailResponseFormat): Promise<GmailRes.GmailThread> => Google.gmailCall(acctEmail, 'GET', `threads/${threadId}`, {
-      format,
-    }),
+    threadGet: (acctEmail: string, threadId: string, format?: GmailResponseFormat, progressCb?: ProgressCb): Promise<GmailRes.GmailThread> =>
+      Google.gmailCall(acctEmail, 'GET', `threads/${threadId}`, {
+        format,
+      }, { download: progressCb }),
```
[the remark this answers is not in the document]
Comment (michael-volynets) [reply extension/js/common/api/google.ts]: I was testing it and it took almost 30% percent of whole the loading time.

**Comment by tomholub** [comment]
Approved

**Comment by michael-volynets** [comment]
So then I can switch out on the new task? Or shall I start trying to write some tests?

**Comment by tomholub** [comment]
Actually, since you mentioned it. This would be a good opportunity to try and write some tests. I will test and merge this PR. You can make an issue to create tests in another PR.

**Comment by tomholub** [comment]
You will want to craft a composer window URL that includes the thread of some other message in it, so that it attempts to reply to that message. Then you test behavior - three dots will load, once loaded, clicking three dots causes text to appear. Do this once for an encrypted message (standard), once for a pgp/mime message, and once for a plain message. That way you'll end up with three tests. This would fall under `compose.ts` tests.

Remember, do not use id= or class= selectors for tests. Each testable element has a special tag attribute for tests, eg `data-test="action-expand-quoted-text"`, so you'll need to add these into the markup.

**Comment by tomholub** [comment]
@michael-volynets I need one change - pulling the message should not block user from starting to type their message.

I mentioned it here, but unfortunately this info gets lost easily in the review comments:

> So you may need to add a loading % sign, maybe next to the three dots, eg "Loading previous message quote... 18%". That means the three dots should probably be visible while this happens, but have opacity 50% and clicking it should not do anything (maybe should also have cursor: default). Then when it loads, opacity of (...) goes to 1 and it can be clicked if user wants to.

image

For the screenshot above I used `opacity: 0.6` while it's loading.

**Comment by tomholub** [comment]
The reason is that as is, it may cause a degraded user experience. I want to start typing my reply, not watch a loader (if the message was 20mb big, I may be staring at the loader for a while). If you do it as in the screenshot above, you're allowing the user to do their work while this is loading.

That means you'll also have to resolve a situation when user hits send but it's still loading. You can add a check for this in `throwIfFormNotReady`

**Comment by michael-volynets** [comment]
yes sorry I did not notice that. I agree with you. And I like the way you came up with.
I think it will be a quick change but I have little reworked forwarding messages.
Now when you forward a message, it also shows a loader before the forwarded message is loaded and I can not do the same thing with forwarding messages because there are no three points.

I think would be better if the solution will be more general? I mean that the same code will work with Forwarding and replying messages.
Maybe we can show the loader inside the reply window but not inside three dots container.

**Comment by tomholub** [comment]
I'd the three dots and the loader for forwarded message too, the same way. When done loading (forward), just auto-click it.

**Comment by tomholub** [comment]
Eg the logic is the same, except the auto-click (alternatively don't register a click handler at all and just do the actual append, whatever is easier in code.

**Comment by michael-volynets** [comment]
Okay, got it, I will do it

File: extension/js/common/composer.ts [hunk 23]
```diff
@@ -625,6 +624,9 @@ export class Composer {
     if (String(this.S.cached('input_to').val()).length) { // evaluate any recipient errors earlier treated as gentle
       await this.parseRenderRecipients('harshRecipientErrs');
     }
+    if (this.S.cached('icon_show_prev_msg').hasClass('progress')) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Good. I tend to save state in the DOM sometimes. I know the React community and others consider it a bad practice, but it has worked for me. Thanks for following the style of surrounding code!

File: extension/js/common/composer.ts [hunk 24]
```diff
@@ -1409,7 +1382,9 @@ export class Composer {
               kisWithPp: await Store.keysGetAllWithPassphrases(this.urlParams.acctEmail),
               encryptedData: Buf.fromUtfStr(stringContent)
             });
-            this.renderGettingPreviousMessageProgress(60 + (40 / encryptedCount) * (index + 1));
+            if (progressCb) {
+              progressCb(60 + (40 / encryptedCount) * (index + 1));
+            }
             if (decrptResult.success) {
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Typo :) `decrptResult`

File: extension/js/common/composer.ts [hunk 25]
```diff
@@ -625,6 +624,9 @@ export class Composer {
     if (String(this.S.cached('input_to').val()).length) { // evaluate any recipient errors earlier treated as gentle
       await this.parseRenderRecipients('harshRecipientErrs');
     }
+    if (this.S.cached('icon_show_prev_msg').hasClass('progress')) {
+      throw new ComposerNotReadyError('Retrieving previous message, please wait.');
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Whoops I deleted the previous comment. Thanks for following the surrounding code style, even if checking state from DOM is not considered the best practice, it has worked for me.

**Comment by tomholub** [comment]
Actually other than the typo, code is good. I'll go ahead and test it.

**Comment by tomholub** [comment]
There is one gotcha to cover - missing pass phrase.

Go to FlowCrypt Settings -> Security -> Require pass phrase.

Then try to reply to a message. I'll cover this one since I already have the code open.

**Comment by tomholub** [comment]
I'm just fixing a few details, hard to say if I broke it by editing the code or it was already that way.

reply -> don't manual expand -> write something -> safe draft -> reopen draft (missing newlines):

image

**Comment by michael-volynets** [comment]
Okay, Thanks. 
Yes, I forgot to about passphrases. If you haven't covered it yet I can continue to cover that

**Comment by tomholub** [comment]
It's done, you can review the commits above

**Comment by tomholub** [comment]
No worries, the pass phrases are easy to forget that they may be missing.

One of my test browsers has settings to forget pass phrase (forgot if it's FF or chrome), so that I would bump into it sometimes as I test.

**Comment by tomholub** [comment]
For the last commit https://github.com/FlowCrypt/flowcrypt-browser/pull/1791/commits/cac7315334b7f3e2d225b3035f35ae811d920258 I have decided to turn it off completely when loading from a draft.

The reason is that comparing our own version of `this.msgExpandingHTMLPart` with `this.S.cached('input_text').html()` is likely to cause unexpected behavior. For example, once you insert HTML into the dom, it's up to the browser to manipulate it any way it wants. Eg what if the browser normalizes `<br>` tags to `<br/>` or the other way around etc.. it's not really possible to compare strings that way reliably once they touched the dom.

Instead I make the loaded draft just render and include the quote, which seems like a decent UX compromise for drafts.

**Comment by tomholub** [comment]
Also missed an important things (but maybe I didn't mention it) - to actually include the quoted part in the sent message :laughing: I'm pushing a commit soon.

**Comment by tomholub** [comment]
Also updating so that the `On...` header is part of quoted content in `pgp_block`

image

Before the change it would still show as part of the response.

**Comment by tomholub** [comment]
Ok I think I'm done. @michael-volynets please review my code in case you spot any issues or have comments.

**Comment by tomholub** [comment]
You can review a range of commits (eg only review my changes instead of re-reviewing everything) by clicking the filter:

image

Then shift + clicking my commits.

**Comment by tomholub** [comment]
Tom +3 hours testing & improvements & polishing

**Comment by michael-volynets** [comment]
> Also missed an important thing (but maybe I didn't mention it) - to actually include the quoted part in the sent message  I'm pushing a commit soon.

Ahahah  :laughing:, I think I was so excited working on the other UI things so forgot the main goal of the task.

I've reviewed the code you wrote and I didn't find any issues.
But there is one minor thing:
`private decryptPreviousMsg = async (index: number, encryptedCount: number, encryptedData: Buf, progressCb?: ProgressCb): Promise<string>`, 
I don't think that's a great idea to create the function containing these parameters `index, encryptedCount` and maybe `progressCb` because it isn't necessary, I would leave progress all the calculations of progress in `getAndDecryptPreviousMessage` function.

If you confirm that we need to change it I can change it now because it's opened in my VScode right now 

Also, there are some things I have learnt from your code 
Thanks for the help.

**Comment by tomholub** [comment]
> and maybe `progressCb` because it isn't necessary, I

Thanks for the review. I didn't 100% understand the progress calculation, so I just copy/pasted and it shows. That make sense, please update it.

**Comment by michael-volynets** [comment]
I've changed it
